src/dbus_api/pool.rs

```diff
@@ -149,6 +143,52 @@ fn destroy_filesystems(m: &MethodInfo<MTFn<TData>, TData>) -> MethodResult {
     Ok(vec![msg])
 }
 
+fn snapshot_filesystem(m: &MethodInfo<MTFn<TData>, TData>) -> MethodResult {
+    let message: &Message = m.msg;
+    let mut iter = message.iter_init();
+
+    let filesystem: dbus::Path<'static> = get_next_arg(&mut iter, 0)?;
+    let snapshot_name: &str = get_next_arg(&mut iter, 0)?;
+
+    let dbus_context = m.tree.get_data();
+    let object_path = m.path.get_name();
+    let return_message = message.method_return();
+    let default_return = MessageItem::ObjectPath(default_object_path());
+
+    let pool_path = m.tree
+        .get(object_path)
+        .expect("implicit argument must be in tree");
+    let pool_uuid = get_data!(pool_path; default_return; return_message).uuid;
+
+    let fs_uuid = match m.tree.get(&filesystem) {
+        Some(op) => get_data!(op; default_return; return_message).uuid,
+        None => {
+            let message = format!("no data for object path {}", filesystem);
+            let (rc, rs) = code_to_message_items(DbusErrorEnum::INTERNAL_ERROR, message);
```

Comment:
I don't think this is an internal error, NOTFOUND is probably better. It's easy enough for the client to specify an obsolete object path. They call GetManagedObjects, identify the filesystem object path of interest, and then call snapshot. By that time, the filesystem may have gone away.